crates/viewer/re_time_panel/src/data_density_graph.rs

```diff
@@ -366,6 +374,253 @@ fn smooth(density: &[f32]) -> Vec<f32> {
+        },
+    );
+
+    for (chunk, time_range, num_events) in chunk_ranges {
+        add_data_point(chunk, time_range, num_events);
+    }
+
+    let hovered_x_range = (time_ranges_ui
+        .x_from_time_f32(hovered_time_range.min().into())
+        .unwrap_or(f32::MAX)
+        - MARGIN_X)
+        ..=(time_ranges_ui
+            .x_from_time_f32(hovered_time_range.max().into())
+            .unwrap_or(f32::MIN)
+            + MARGIN_X);
+
+    density_graph.buckets = smooth(&density_graph.buckets);
+
+    density_graph.paint(
+        data_density_graph_painter,
+        row_rect.y_range(),
+        time_area_painter,
+        graph_color(ctx, &item.to_item(), ui),
+        hovered_x_range.clone(),
+    );
+
+    if num_hovered_messages > 0 {
+        ctx.selection_state().set_hovered(item.to_item());
+
+        if time_area_response.clicked_by(egui::PointerButton::Primary) {
+            ctx.selection_state().set_selection(item.to_item());
+            time_ctrl.set_time(hovered_time_range.min());
+            time_ctrl.pause();
+        } else if ui.ctx().dragged_id().is_none() {
+            egui::show_tooltip_at_pointer(
+                ui.ctx(),
+                ui.layer_id(),
+                egui::Id::new("data_tooltip"),
+                |ui| {
+                    show_row_ids_tooltip(
+                        ctx,
+                        ui,
+                        time_ctrl,
+                        db,
+                        item,
+                        hovered_time_range,
+                        num_hovered_messages,
+                    );
+                },
+            );
+        }
+    }
+}
+
+// if sorted:
+// - binary search, show hovered things
+// - if its small enough, full scan
+// - otherwise, hover full chunk
+
+fn visit_chunk_sub_range(
+    chunk: &Chunk,
+    component_name: Option<ComponentName>,
+    timeline: Timeline,
+    time_range: ResolvedTimeRange,
+    mut visitor: impl FnMut(Arc<Chunk>, &ChunkTimeline, usize),
+) {
+    let query = RangeQuery::new(timeline, time_range);
+    if let Some(component_name) = component_name {
+        let chunk = Arc::new(chunk.range(&query, component_name));
+        let Some(num_events) = chunk.num_events_for_component(component_name) else {
+            return;
+        };
+        let Some(chunk_timeline) = chunk.timelines().get(&timeline) else {
+            return;
+        };
+        visitor(Arc::clone(&chunk), chunk_timeline, num_events);
+    } else {
+        for component_name in chunk.component_names() {
+            let chunk = Arc::new(chunk.range(&query, component_name));
+            let Some(num_events) = chunk.num_events_for_component(component_name) else {
+                continue;
+            };
+            let Some(chunk_timeline) = chunk.timelines().get(&timeline) else {
+                continue;
+            };
+            visitor(Arc::clone(&chunk), chunk_timeline, num_events);
+        }
```

Comment:
I think this should be turned into a `chunk.range_entity(&query)` or something, that also de-duplicates the cunks.

Maybe @teh-cmc has some thoughts here

Reply:
Actually, I think we want to use the query cache… once it is fully built :)

Reply:
Actually no - we change the view too often for a cache to be helpful.

I think a better approach is this:
- if it is a small chunk, do a linear scan
- if it is a big chunk that is sorted on the timeline, to a cheap binary-search
- if it is a big chunk that is unsorted, then return the whole chunk (bad picking behavior, but performance is more important)

Hopefully we shouldn't have large unsorted chunks, and the compaction code should make sure of that, because large unsorted chunks will always be bad for performance